src/widget_type.h

```diff
@@ -701,7 +701,7 @@ class Scrollbar {
 	 * Gets the number of elements in the list
 	 * @return the number of elements
 	 */
-	inline uint16_t GetCount() const
+	inline int GetCount() const
```

Comment:
random question I don't know the answer to:

do we want to use `int32_t` and `INT32_MAX`, instead of `int` and `INT_MAX`, just to have it more clear to everyone what type we talk about exactly?

An `int` for a for-loop is ofc perfectly fine, but for things like this, I wonder if we should be more explicit. Open question, don't know the answer myself :)

Reply:
That it is `int` is because of one of the properties of `int` that `int32_t` does not necessarily have. It's basically the default type things get promoted to. 

Going with `int` in the for-loops will eventually get you into problems if you need to convert it back to `int32_t` when some compiler makes `int` 64 bits. So, in my opinion, it's basically an all-or-nothing approach that should be taken to just disallow `int`/`unsigned int`/`uint`. That's definitely not in scope for this PR.

Reply:
This is one of the reasons I've slowly been working on things like `GetScrolledItemFromWidget()` where the caller doesn't have to do anything with different integer types.

Reply:
@rubidium42 sometimes it feels we talk about different dimensions :) Let me retry this ..

My for-loop remark was to avoid the common response: OMG HE WANTS TO CHANGE ALL `int` TO `int32_t`. I am not. A `for (int i = 0; i < 10; i++)` is perfectly fine in my world.

What I am questioning however, if we shouldn't be more explicit in our datatypes in structs and these kind of functions. `int32_t` is clear to everyone what is meant, where `int` is not always all that clear. And when someone would compile it on a 16bit compiler, would give different results ;)

As `int` is an `int32_t` on all compilers we care about, "promotion" isn't an issue, as they are identical in type. So the code would compile exactly the same with both types as far as I know.

So my only question is: for this PR, for the changes to `int`: shouldn't we be doing `int32_t` instead, to just make people's life a bit easier that we actually mean a 32bit signed value?

And this is just an honest question, mostly to avoid people starting to use `int` and `uint` in all sorts of data containers we have. Do we actually like / want that?

Reply:
And maybe to be more verbose in what I mean:

https://github.com/TrueBrain/OpenTTD/pull/28

Same result, just more explicit that we want a 32bit signed integer for our scroll location.

Reply:
The code interfacing with this does use `int` quite often. For example `Window::GetRowFromWidget` which returns `int` and `INT_MAX`.
In my mind it would be conceptually weird to let that generate `INT_MAX` and then later check against `INT32_MAX`, so that should be changed to keep the types and constant the same... even though they happen to be the same type/value with our current compilers.

How I understood your comment about `int` in for-loops being fine was that you see no problem in:
```
int32_t start_value = 32;
int32_t end_value = 64;
for (int i = start_value; i < end_value; ++i) something(i);
```
That, in my mind, is just messy.

Finally, I've got the feeling such debates have happened in the past and just removing `int` and `uint` from the codebase completely would remove such debates in the future and just set a clear standard that everyone can abide by instead of asking open questions every now and then whether something should be `int` or `int32_t`.

Reply:
The reason we keep having certain debates, is because we are getting cleaner and cleaner code in terms of typing etc. But it is a slow process. But we keep pushing it forward, which has made the code a lot more readable over the last few years. These conversations are, how ever annoying, also healthy. Just so we deliberately make choices :)

I do not remember we ever talked about flat-out removing `int`, but I do remember we have had talks about making storage-types more explicit, and not depend on the "at least NNN" of `int` and friends.

Which brought me to my original question :) If we deliberately say: no, we use `int` here, I am fine with it. As long as we have a good reason for it, and not "because the old us did it, so I am going to do it". That way we never grow better, in our code base :) But you have been pushing most of these "moving forward", so I am preaching to the choir here :)

Reply:
Yes, it should eventually probably be `int32_t`. However, this is likely going to be backported so I'm also trying to keep it relatively small while not mix-and-matching types.

Changing the types to `int32_t` would, in my opinion, also mean changing `GetScrolledRowFromWidget` to return `int32_t` and `INT32_MAX` in 'wrong position' situations. By proxy that should probably also affect `GetRowFromWidget` in the same way to keep them consistent, or keep the `pos` in `GetScrolledRowFromWidget` as `int` and check that against `INT_MAX`. Next to that all users of the scrollbar (and potentially) `GetRowFromWidget` should be checked and amended.

Or are we fine with only changing `int` to `int32_t` when we (have to) change the line?

Reply:
May I suggest a typedef (in `using x = y;`) format to define the type, and a constexpr of `std::numeric_limits<x>::max()` for the max value?

Keep it only to scrollbar for now for this PR for back-portability, and then later it we can update other code that uses it if we decide it needs it?